infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/runtime/LogReplicationClientServerRouter.java

```diff
@@ -0,0 +1,613 @@
+                    message.getPayload().getPayloadCase(),
+                    t.getClass().getSimpleName(),
+                    t.getMessage(),
+                    t);
+        }
+    }
+
+    public boolean isConnectionInitiator(LogReplicationSession session) {
+        return sessionManager.isConnectionStarterForSession(session);
+    }
+
+    private void removeSessionInfo(LogReplicationSession session) {
+        sessionToSinkVerifyLeadership.remove(session);
+        sessionToOutstandingRequests.remove(session);
+        sessionToRuntimeFSM.remove(session);
+        sessionToRemoteClusterDescriptor.remove(session);
+        sessionToRequestIdCounter.remove(session);
+        sessionToLeaderConnectionFuture.remove(session);
+    }
+
+    public void shutDownMsgHandlerServer() {
+        // A executor service to create the shutdown threads plus name the threads correctly.
+        final ExecutorService shutdownService = Executors.newSingleThreadExecutor(
+                new ServerThreadFactory("ReplicationCorfuServer-shutdown-",
+                        new ServerThreadFactory.ExceptionHandler()));
+
+        // Turn into a list of futures on the shutdown, returning
+        // generating a log message to inform of the result.
+        CompletableFuture shutdownFuture = CompletableFuture.runAsync(() -> {
+            try {
+                log.info("Shutting down {}", msgHandler.getClass().getSimpleName());
+                msgHandler.shutdown();
+                log.info("Cleanly shutdown {}", msgHandler.getClass().getSimpleName());
+            } catch (Exception e) {
+                log.error("Failed to cleanly shutdown {}", msgHandler.getClass().getSimpleName(), e);
+            }
+        }, shutdownService);
+
+        CompletableFuture.allOf(shutdownFuture).join();
+        shutdownService.shutdown();
+    }
+
+    private boolean isOutgoingSession(LogReplicationSession session) {
+        return sessionManager.getOutgoingSessions().contains(session);
+    }
+
+    /**
+     * Channel Adapter On Error Callback
+     */
+    public synchronized void onError(Throwable t, LogReplicationSession session) {
+        sessionToRuntimeFSM.get(session).input(new LogReplicationRuntimeEvent(LogReplicationRuntimeEvent.LogReplicationRuntimeEventType.ERROR, t));
+    }
+
+
+    /**
+     * Verify Message is of valid Log Replication type.
+     */
+    private boolean isValidMessage(CorfuMessage.RequestPayloadMsg message) {
+        return message.getPayloadCase().equals(CorfuMessage.RequestPayloadMsg.PayloadCase.LR_ENTRY) ||
+                message.getPayloadCase().equals(CorfuMessage.RequestPayloadMsg.PayloadCase.LR_METADATA_REQUEST) ||
+                message.getPayloadCase().equals(CorfuMessage.RequestPayloadMsg.PayloadCase.LR_LEADERSHIP_QUERY);
+    }
+
+    private void startReplication(LogReplicationSession session, String nodeId) {
+        replicationManager.startLogReplicationRuntime(session);
```

Comment:
this one starts the source FSM.
We create the FSM when we discover an outgoing session, but we want the FSM to be started only after a connection has been established.

Reply:
since this comment is related, https://github.com/CorfuDB/CorfuDB/pull/3560#discussion_r1152616351

I agree replicationManager need not be stored in router class, removing it but just pointing out that `startLogReplicationRuntime` will now reside in router.

Reply:
is there an advantage of starting the fsm when a connection has been established?  Even if we start it earlier, it will wait until an incoming event is received, right?

Reply:
This has been changed to what it was previously...i.e., the FSM.start() is called as soon as it is created.
The on-connection up event actually starts the FSM